controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -339,12 +339,32 @@ func (r *BareMetalHostReconciler) credentialsErrorResult(err error, request ctrl
 }
 
 // hasRebootAnnotation checks for existence of reboot annotations and returns true if at least one exist
-func hasRebootAnnotation(host *metal3v1alpha1.BareMetalHost) bool {
-	for annotation := range host.Annotations {
+func hasRebootAnnotation(info *reconcileInfo) (hasReboot, hardMode bool) {
+	for annotation, value := range info.host.GetAnnotations() {
+		// Don't use a break here as we may have multiple clients setting
+		// reboot annotations and we always want hard requests honoured
 		if isRebootAnnotation(annotation) {
-			return true
+			hasReboot = true
+			if isHardReboot(value, info) {
+				hardMode = true
+			}
 		}
 	}
+	return
+}
+
+// isHardReboot checks if reboot annotation requests a hard reboot and returns true if so
+func isHardReboot(annotation string, info *reconcileInfo) bool {
+	annotation_map := map[string]string{}
+	err := json.Unmarshal([]byte(annotation), &annotation_map)
+	if err != nil {
+		info.publishEvent("InvalidAnnotationValue", fmt.Sprintf("could not parse reboot annotation (%s), assuming soft reboot mode", err))
+		info.log.Info("Could not parse reboot annotation value: " + err.Error())
```

Comment:
It would be helpful to add more details to the log message - the bmh name and the value that failed to be parsed.
I'd also add the "assuming soft reboot mode" text to the log message as well.

Reply:
I think the logger already includes the BMH name, so we shouldn't need to explicitly add it to the event/log string.

Agree adding the value would be helpful in the logs, otherwise we just get something like `invalid character '}' after object key`

Also we're saying `reboot annotation (%s)` but putting err in the string for the event - is that what we want, or should that show the value we found to be invalid, e.g `could not parse reboot annotation value ("{\"thisisabadvalue\"}") - invalid json, assuming soft-reboot`

Reply:
Thanks, taken care of I believe :)